src/schema/posts.ts

```diff
@@ -1550,6 +1667,112 @@ export const resolvers: IResolvers<any, Context> = {
       });
       return { _: true };
     },
+    votePost: async (
+      source,
+      { id, vote }: { id: string; vote: UserPostVote },
+      ctx: Context,
+    ): Promise<GQLEmptyResponse> => {
+      try {
+        if (!Object.values(UserPostVote).includes(vote)) {
+          throw new ValidationError('Unsupported vote type');
+        }
+
+        const post = await ctx.con.getRepository(Post).findOneByOrFail({ id });
+        await ensureSourcePermissions(ctx, post.sourceId);
+        const userPostRepo = ctx.con.getRepository(UserPost);
+
+        switch (vote) {
+          case UserPostVote.Up:
+            await userPostRepo.save({
+              postId: id,
+              userId: ctx.userId,
+              vote: UserPostVote.Up,
+              hidden: false,
+            });
+
+            break;
+          case UserPostVote.Down:
+            await userPostRepo.save({
+              postId: id,
+              userId: ctx.userId,
+              vote: UserPostVote.Down,
+              hidden: true,
+            });
+
+            break;
+          case UserPostVote.None:
+            await userPostRepo.save({
+              postId: id,
+              userId: ctx.userId,
+              vote: UserPostVote.None,
+              hidden: false,
+            });
+
+            break;
+          default:
+            throw new ValidationError('Unsupported vote type');
+        }
+      } catch (err) {
+        // Foreign key violation
+        if (err?.code === TypeOrmError.FOREIGN_KEY) {
+          throw new NotFoundError('Post or user not found');
+        }
+
+        throw err;
+      }
+
+      return { _: true };
+    },
+    dismissPostFeedback: async (
+      source,
+      { id }: { id: string },
+      ctx: Context,
+    ) => {
+      try {
+        const post = await ctx.con.getRepository(Post).findOneByOrFail({ id });
+        await ensureSourcePermissions(ctx, post.sourceId);
+
+        await ctx.con.transaction(async (entityManager) => {
+          const userPost = await entityManager
+            .getRepository(UserPost)
+            .findOneBy({
+              postId: id,
+              userId: ctx.userId,
+            });
+
+          if (userPost) {
```

Comment:
If you want we can do it in one query, not highly important for this endpoint but more for learning purposes:
https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/
insert on conflict ftw!

Reply:
Yeah I tried `repo.upsert` from `typeorm` but would still not work because it does not support passing `UPDATE` statement for column like `repo.update`. I would have to do this in raw SQL which I thought is worse so I went with above. Is there any way that you know it can be done through typeorm?

Reply:
Mentioned it here https://github.com/dailydotdev/daily-api/pull/1389#discussion_r1291021136

Reply:
You can do it with typeorm query builder, not the actual orm methods. You can look for `onConflict` in our codebase, I believe there are few examples

Reply:
@idoshamun thanks 👌 implemented in 1afbd774c2790040af8b37c7d7b7cd4c514634ee

One note, `onConflict` seems to be deprecated in `typeorm` https://github.com/typeorm/typeorm/issues/8168 and proposed `orUpdate` replacement does not support our case since it only allows to specify which columns will be overwritten in `UPDATE` query.

So I am thinking we might just leave it with my previous implementation without using `onConflict` which will be removed in the future versions. WDYT?

Reply:
Sounds good to me. Your call :)
Thanks for checking